jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java

```diff
@@ -0,0 +1,460 @@
+     * When an entry can't be found in the cache, the global list is iterated
+     * normally so the cache has no visible effect besides performance.
+     */
+    private final ThreadLocal<List<Entry>> cache;
+    private final Lock reservationLock = new ReentrantLock();
+    private final int maxEntries;
+    private final int cacheSize;
+    private volatile boolean closed;
+    private volatile int maxMultiplex = 1;
+    private volatile int maxUsageCount = -1;
+
+    /**
+     * Construct a Pool with the specified thread-local cache size.
+     *
+     * @param maxEntries the maximum amount of entries that the pool will accept.
+     * @param cacheSize the thread-local cache size. A value less than 1 means the cache is disabled.
+     */
+    public Pool(int maxEntries, int cacheSize)
+    {
+        this.maxEntries = maxEntries;
+        this.cacheSize = cacheSize;
+        if (cacheSize > 0)
+            this.cache = ThreadLocal.withInitial(() -> new ArrayList<Entry>(cacheSize));
+        else
+            this.cache = null;
+    }
+
+    public int getMaxEntries()
+    {
+        return maxEntries;
+    }
+
+    public int getMaxMultiplex()
+    {
+        return maxMultiplex;
+    }
+
+    public final void setMaxMultiplex(int maxMultiplex)
+    {
+        if (maxMultiplex < 1)
+            throw new IllegalArgumentException("Max multiplex must be >= 1");
+        this.maxMultiplex = maxMultiplex;
+    }
+
+    public int getMaxUsageCount()
+    {
+        return maxUsageCount;
+    }
+
+    public final void setMaxUsageCount(int maxUsageCount)
+    {
+        if (maxUsageCount == 0)
+            throw new IllegalArgumentException("Max usage count must be != 0");
+        this.maxUsageCount = maxUsageCount;
+    }
+
+    /**
+     * Create a new disabled slot into the pool. The returned entry
+     * won't be acquirable as long as {@link Entry#enable(Object)}
+     * has not been called.
+     *
+     * @param maxReservations the max desired number of reserved entries,
+     * or a negative number to always trigger the reservation of a new entry.
+     * @return a disabled entry that is contained in the pool,
+     * or null if the pool is closed or if the pool already contains
+     * {@link #getMaxEntries()} entries.
+     */
+    public Entry reserve(int maxReservations)
+    {
+        reservationLock.lock();
+        try
+        {
+            if (maxReservations >= 0)
+            {
+                long pending = sharedList.stream().filter(entry -> entry.getPooled() == null).count();
+                if (pending >= maxReservations)
+                    return null;
+            }
+
+            if (!closed && sharedList.size() < maxEntries)
+            {
+                Entry entry = new Entry();
+                sharedList.add(entry);
+                return entry;
+            }
+            return null;
+        }
+        finally
+        {
+            reservationLock.unlock();
+        }
```

Comment:
As discussed, I think the lock can be made smaller (just rename it from `reservationLock` to `lock` as it's also used to protect `closed`).

Only 138-142 needs to be guarded by the lock.